File: app/Models/MrsoolDelivery.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

/**
 * One courier request to Mrsool (مرسول) for a Zooboxi express order.
 *
 * Append-only ledger: an order may have several rows (a failed/cancelled
 * attempt followed by a retry), but at most ONE non-terminal row at a time.
 */
class MrsoolDelivery extends Model
{
    use HasFactory;

    protected $table = 'mrsool_deliveries';

    protected $guarded = [];

    protected $casts = [
        'mrsool_order_id' => 'integer',
        'is_partial'      => 'boolean',
        'price_quote'     => 'float',
        'courier_lat'     => 'float',
        'courier_lng'     => 'float',
        'events'          => 'array',
        'pickup_images'   => 'array',
        'dropoff_images'  => 'array',
        'raw_last'        => 'array',
        'requested_at'    => 'datetime',
        'assigned_at'     => 'datetime',
        'picked_up_at'    => 'datetime',
        'delivered_at'    => 'datetime',
        'failed_at'       => 'datetime',
        'last_synced_at'  => 'datetime',
    ];

    // ─── Phases (our own lifecycle, derived from the Mrsool status) ──

    const PHASE_SEARCHING  = 'searching';
    const PHASE_ASSIGNED   = 'assigned';
    const PHASE_IN_TRANSIT = 'in_transit';
    const PHASE_DELIVERED  = 'delivered';
    const PHASE_FAILED     = 'failed';

    const TERMINAL_PHASES = [self::PHASE_DELIVERED, self::PHASE_FAILED];

    // ─── Mrsool statuses (exact API strings) ────────────────────

    const S_COURIER_PENDING      = 'COURIER_PENDING';
    const S_COURIER_ASSIGNED     = 'COURIER_ASSIGNED';
    const S_COURIER_REASSIGNED   = 'COURIER_REASSIGNED';
    const S_PICKUP_ARRIVED       = 'PICKUP_ARRIVED';
    const S_COLLECTING           = 'COLLECTING';
    const S_CONFIRMED_PICKUP     = 'CONFIRMED_PICKUP';
    const S_WAITING_FOR_DELIVERY = 'WAITING_FOR_DELIVERY';
    const S_DELIVERING           = 'DELIVERING';
    const S_DROPOFF_ARRIVED      = 'DROPOFF_ARRIVED';
    const S_PARTIALLY_DELIVERED  = 'PARTIALLY_DELIVERED';
    const S_DELIVERED            = 'DELIVERED';
    const S_RETURN               = 'RETURN';
    const S_CANCELED             = 'CANCELED';
    const S_EXPIRED              = 'EXPIRED';

    /**
     * The ONE mapping from a Mrsool status to our phase. Anything unknown keeps
     * the delivery in its current phase (we never guess a terminal state).
     */
    const STATUS_PHASES = [
        self::S_COURIER_PENDING      => self::PHASE_SEARCHING,
        self::S_COURIER_ASSIGNED     => self::PHASE_ASSIGNED,
        self::S_COURIER_REASSIGNED   => self::PHASE_ASSIGNED,
        self::S_PICKUP_ARRIVED       => self::PHASE_ASSIGNED,
        self::S_COLLECTING           => self::PHASE_ASSIGNED,
        self::S_CONFIRMED_PICKUP     => self::PHASE_IN_TRANSIT,
        self::S_WAITING_FOR_DELIVERY => self::PHASE_IN_TRANSIT,
        self::S_DELIVERING           => self::PHASE_IN_TRANSIT,
        self::S_DROPOFF_ARRIVED      => self::PHASE_IN_TRANSIT,
        self::S_DELIVERED            => self::PHASE_DELIVERED,
        self::S_PARTIALLY_DELIVERED  => self::PHASE_DELIVERED,
        self::S_RETURN               => self::PHASE_FAILED,
        self::S_CANCELED             => self::PHASE_FAILED,
        self::S_EXPIRED              => self::PHASE_FAILED,
    ];

    /** Arabic labels for the 14 Mrsool statuses (shown in the app + Filament). */
    const STATUS_LABELS = [
        self::S_COURIER_PENDING      => 'جارٍ البحث عن مندوب',
        self::S_COURIER_ASSIGNED     => 'تم تعيين مندوب',
        self::S_COURIER_REASSIGNED   => 'تم تغيير المندوب',
        self::S_PICKUP_ARRIVED       => 'المندوب وصل الفرع',
        self::S_COLLECTING           => 'جارٍ استلام الطلب',
        self::S_CONFIRMED_PICKUP     => 'تم استلام الطلب',
        self::S_WAITING_FOR_DELIVERY => 'بانتظار التوصيل',
        self::S_DELIVERING           => 'في الطريق للعميل',
        self::S_DROPOFF_ARRIVED      => 'المندوب وصل للعميل',
        self::S_PARTIALLY_DELIVERED  => 'تم التوصيل جزئياً',
        self::S_DELIVERED            => 'تم التوصيل',
        self::S_RETURN               => 'مرتجع',
        self::S_CANCELED             => 'ملغي',
        self::S_EXPIRED              => 'منتهي (لم يُعثر على مندوب)',
    ];

    // ─── Relationships ──────────────────────────────────────────

    public function order()
    {
        return $this->belongsTo(ZooboxiOrder::class, 'zooboxi_order_id');
    }

    public function requestedBy()
    {
        return $this->belongsTo(User::class, 'requested_by');
    }

    // ─── Scopes ─────────────────────────────────────────────────

    /** Non-terminal deliveries — the ones the poller and the app care about. */
    public function scopeActive(Builder $query): Builder
    {
        return $query->whereNotIn('phase', self::TERMINAL_PHASES);
    }

    // ─── Helpers ────────────────────────────────────────────────

    public function isTerminal(): bool
    {
        return in_array($this->phase, self::TERMINAL_PHASES, true);
    }

    /** The branch may only pull the request back before the courier collects. */
    public function canCancel(): bool
    {
        return in_array($this->phase, [self::PHASE_SEARCHING, self::PHASE_ASSIGNED], true);
    }

    public function getStatusLabelAttribute(): ?string
    {
        return self::STATUS_LABELS[$this->status] ?? $this->status;
    }

    /** Map a raw Mrsool status to our phase (null when unknown). */
    public static function phaseFor(?string $status): ?string
    {
        return self::STATUS_PHASES[strtoupper((string) $status)] ?? null;
    }

    public static function labelFor(?string $status): ?string
    {
        if ($status === null || $status === '') {
            return null;
        }
        return self::STATUS_LABELS[strtoupper($status)] ?? $status;
    }
}

File: exhibition_manager_app/lib/features/zooboxi_orders/data/models/mrsool_delivery.dart
// Models for the Mrsool (مرسول) express last-mile leg of a Zooboxi order
// (GET/POST /zooboxi-orders/{id}/mrsool*).
//
// Every field is parsed defensively — the backend omits nulls and the Mrsool
// payload itself is only partially populated until a courier is assigned.

import 'package:exhibition_manager_app/core/localization/app_localizations.dart';
import 'package:exhibition_manager_app/shared/utils/date_names.dart';

/// Mrsool delivery phases (backend-mapped, the single source of truth).
class MrsoolPhase {
  MrsoolPhase._();

  static const String searching = 'searching';
  static const String assigned = 'assigned';
  static const String inTransit = 'in_transit';
  static const String delivered = 'delivered';
  static const String failed = 'failed';
}

/// The courier assigned to a delivery (empty until Mrsool assigns one).
class MrsoolCourier {
  final String? name;
  final String? phone;
  final double? latitude;
  final double? longitude;

  const MrsoolCourier({this.name, this.phone, this.latitude, this.longitude});

  static MrsoolCourier? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    final courier = MrsoolCourier(
      name: _str(json['name']),
      phone: _str(json['phone']),
      latitude: _nullableDouble(json['lat'] ?? json['latitude']),
      longitude: _nullableDouble(json['lng'] ?? json['longitude']),
    );
    return courier.isEmpty ? null : courier;
  }

  bool get isEmpty => name == null && phone == null && latitude == null && longitude == null;
  bool get hasName => name != null && name!.isNotEmpty;
  bool get hasPhone => phone != null && phone!.isNotEmpty;
  bool get hasLocation => latitude != null && longitude != null;
}

/// One entry of the Mrsool `events_history`.
class MrsoolEvent {
  final String event;
  final String? label;
  final DateTime? at;

  const MrsoolEvent({required this.event, this.label, this.at});

  factory MrsoolEvent.fromJson(Map<String, dynamic> json) => MrsoolEvent(
        event: _str(json['event']) ?? '',
        label: _str(json['label']),
        at: _date(json['at'] ?? json['created_at']),
      );

  /// Prefer the app's own translation of the status, then the server label.
  String get displayLabel {
    final key = mrsoolStatusKey(event);
    if (key != null) return AppLocalizations.translate(key);
    if (label != null && label!.isNotEmpty) return label!;
    return event;
  }
}

/// One courier request (one row of `mrsool_deliveries`).
class MrsoolDelivery {
  final int id;
  final int? mrsoolOrderId;
  final String? status; // raw Mrsool status string
  final String? statusLabel; // server-provided Arabic label (fallback)
  final String phase;
  final bool isPartial;
  final double? priceQuote;
  final MrsoolCourier? courier;
  final List<MrsoolEvent> events;
  final List<String> pickupImages;
  final List<String> dropoffImages;
  final String? awbUrl;
  final String? lastError;
  final DateTime? requestedAt;
  final DateTime? assignedAt;
  final DateTime? pickedUpAt;
  final DateTime? deliveredAt;
  final DateTime? failedAt;
  final DateTime? lastSyncedAt;
  final bool canCancel;

  const MrsoolDelivery({
    required this.id,
    this.mrsoolOrderId,
    this.status,
    this.statusLabel,
    this.phase = MrsoolPhase.searching,
    this.isPartial = false,
    this.priceQuote,
    this.courier,
    this.events = const [],
    this.pickupImages = const [],
    this.dropoffImages = const [],
    this.awbUrl,
    this.lastError,
    this.requestedAt,
    this.assignedAt,
    this.pickedUpAt,
    this.deliveredAt,
    this.failedAt,
    this.lastSyncedAt,
    this.canCancel = false,
  });

  factory MrsoolDelivery.fromJson(Map<String, dynamic> json) => MrsoolDelivery(
        id: _int(json['id']),
        mrsoolOrderId: json['mrsool_order_id'] == null ? null : _int(json['mrsool_order_id']),
        status: _str(json['status']),
        statusLabel: _str(json['status_label']),
        phase: _str(json['phase']) ?? MrsoolPhase.searching,
        isPartial: json['is_partial'] == true,
        priceQuote: _nullableDouble(json['price_quote']),
        courier: MrsoolCourier.fromJson(_map(json['courier'])),
        events: _list(json['events'])
            .map((e) => MrsoolEvent.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList(),
        pickupImages: _strings(json['pickup_images']),
        dropoffImages: _strings(json['dropoff_images']),
        awbUrl: _str(json['awb_url']),
        lastError: _str(json['last_error']),
        requestedAt: _date(json['requested_at']),
        assignedAt: _date(json['assigned_at']),
        pickedUpAt: _date(json['picked_up_at']),
        deliveredAt: _date(json['delivered_at']),
        failedAt: _date(json['failed_at']),
        lastSyncedAt: _date(json['last_synced_at']),
        canCancel: json['can_cancel'] == true,
      );

  bool get isSearching => phase == MrsoolPhase.searching;
  bool get isAssigned => phase == MrsoolPhase.assigned;
  bool get isInTransit => phase == MrsoolPhase.inTransit;
  bool get isDelivered => phase == MrsoolPhase.delivered;
  bool get isFailed => phase == MrsoolPhase.failed;

  /// No further updates will ever arrive — stops the auto-refresh timer.
  bool get isTerminal => isDelivered || isFailed;

  /// Localized status headline: our own 14-status map, then the server label.
  String get displayStatus {
    final key = mrsoolStatusKey(status);
    if (key != null) return AppLocalizations.translate(key);
    if (statusLabel != null && statusLabel!.isNotEmpty) return statusLabel!;
    return AppLocalizations.translate(_phaseKey);
  }

  String get _phaseKey => switch (phase) {
        MrsoolPhase.assigned => 'mrsool_status_courier_assigned',
        MrsoolPhase.inTransit => 'mrsool_status_delivering',
        MrsoolPhase.delivered => 'mrsool_status_delivered',
        MrsoolPhase.failed => 'mrsool_failed',
        _ => 'mrsool_status_courier_pending',
      };

  /// The most meaningful timestamp for the current phase.
  DateTime? get phaseAt => switch (phase) {
        MrsoolPhase.delivered => deliveredAt ?? lastSyncedAt,
        MrsoolPhase.failed => failedAt ?? lastSyncedAt,
        MrsoolPhase.inTransit => pickedUpAt ?? assignedAt,
        MrsoolPhase.assigned => assignedAt,
        _ => requestedAt,
      };

  List<String> get allPhotos => [...pickupImages, ...dropoffImages];
}

/// Translation key for one of the 14 Mrsool statuses (null when unknown).
String? mrsoolStatusKey(String? status) {
  if (status == null || status.isEmpty) return null;
  const known = {
    'COURIER_PENDING',
    'COURIER_ASSIGNED',
    'COURIER_REASSIGNED',
    'PICKUP_ARRIVED',
    'COLLECTING',
    'CONFIRMED_PICKUP',
    'WAITING_FOR_DELIVERY',
    'DELIVERING',
    'DROPOFF_ARRIVED',
    'PARTIALLY_DELIVERED',
    'DELIVERED',
    'RETURN',
    'CANCELED',
    'EXPIRED',
  };
  final upper = status.toUpperCase();
  if (!known.contains(upper)) return null;
  return 'mrsool_status_${upper.toLowerCase()}';
}

/// "14:32" today, otherwise "12 يونيو · 14:32" / "12 Jun · 14:32".
String mrsoolTimeLabel(DateTime? value) {
  if (value == null) return '—';
  final d = value.toLocal();
  final time = '${d.hour.toString().padLeft(2, '0')}:${d.minute.toString().padLeft(2, '0')}';
  final now = DateTime.now();
  if (d.year == now.year && d.month == now.month && d.day == now.day) return time;
  return '${d.day} ${monthName(d.month, short: true)} · $time';
}

// ─── Null-safe parsing helpers ────────────────────────────────
String? _str(dynamic v) {
  if (v == null) return null;
  final s = v.toString().trim();
  return s.isEmpty ? null : s;
}

int _int(dynamic v) {
  if (v is int) return v;
  if (v is num) return v.round();
  return int.tryParse('$v') ?? 0;
}

double? _nullableDouble(dynamic v) {
  if (v == null) return null;
  if (v is num) return v.toDouble();
  return double.tryParse(v.toString());
}

DateTime? _date(dynamic v) {
  final s = _str(v);
  return s == null ? null : DateTime.tryParse(s);
}

Map<String, dynamic>? _map(dynamic v) =>
    v is Map ? Map<String, dynamic>.from(v) : null;

List _list(dynamic v) => v is List ? v : const [];

List<String> _strings(dynamic v) =>
    _list(v).map((e) => e?.toString() ?? '').where((e) => e.isNotEmpty).toList();
